arbnode/batch_poster.go

```diff
@@ -393,149 +461,85 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context, batchSeqNum u
-	} else {
-		b.pendingMsgTimestamp = time.Now()
-	}
-	return tx, nil
-}
-
-// Returns the timestamp of the next message to post, or time.Now() if there's no new messages
-func (b *BatchPoster) recomputePendingMsgTimestamp(ctx context.Context, batchCount uint64) error {
-	if batchCount == 0 {
-		// No batches, therefore batch posting was not required
-		b.pendingMsgTimestamp = time.Now()
-		return nil
-	}
-	batchMsgCount, err := b.inbox.GetBatchMessageCount(batchCount - 1)
+	gasLimit, err := b.estimateGas(ctx, sequencerMsg, b.building.segments.delayedMsg)
 	if err != nil {
 		return err
 	}
-	msgCount, err := b.streamer.GetMessageCount()
+	data, err := b.encodeAddBatch(new(big.Int).SetUint64(batchPosition.NextSeqNum), batchPosition.MessageCount, b.building.msgCount, sequencerMsg, b.building.segments.delayedMsg)
 	if err != nil {
 		return err
 	}
-	if msgCount <= batchMsgCount {
-		// There's nothing after the newest batch, therefore batch posting was not required
-		b.pendingMsgTimestamp = time.Now()
-		return nil
+	newMeta := batchPosterPosition{
+		MessageCount:        b.building.msgCount,
+		DelayedMessageCount: b.building.segments.delayedMsg,
+		NextSeqNum:          batchPosition.NextSeqNum + 1,
 	}
-	msg, err := b.streamer.GetMessage(batchMsgCount - 1)
+	err = b.dataPoster.PostTransaction(ctx, nextMessageTime, nonce, newMeta, b.seqInboxAddr, data, gasLimit)
 	if err != nil {
 		return err
 	}
-	b.pendingMsgTimestamp = time.Unix(int64(msg.Message.Header.Timestamp), 0)
+	log.Info(
+		"BatchPoster: batch sent",
+		"sequence nr.", batchPosition.NextSeqNum,
+		"from", batchPosition.MessageCount,
+		"to", b.building.msgCount,
+		"prev delayed", batchPosition.DelayedMessageCount,
+		"current delayed", b.building.segments.delayedMsg,
+		"total segments", len(b.building.segments.rawSegments),
+	)
+	b.building = nil
 	return nil
 }
 
 func (b *BatchPoster) Start(ctxIn context.Context) {
 	b.StopWaiter.Start(ctxIn)
 	b.CallIteratively(func(ctx context.Context) time.Duration {
-		batchSeqNum, err := b.inbox.GetBatchCount()
-		if err != nil {
-			log.Error("error getting inbox batch count", "err", err)
-			return b.config.PostingErrorDelay
-		}
-		if batchSeqNum != b.lastBatchCount {
-			err := b.recomputePendingMsgTimestamp(ctx, batchSeqNum)
-			if err != nil {
-				log.Error("error getting next message time", "err", err)
-				return b.config.PostingErrorDelay
-			}
-			b.lastBatchCount = batchSeqNum
-		}
-		_, err = b.maybePostSequencerBatch(ctx, batchSeqNum)
+		err := b.maybePostSequencerBatch(ctx)
 		if err != nil {
 			b.building = nil
-			log.Error("error posting batch", "err", err)
+			logLevel := log.Error
+			if errors.Is(err, AccumulatorNotFound) {
+				// Likely the inbox tracker just isn't caught up.
+				// Let's see if this error disappears naturally.
+				if b.firstAccErr == (time.Time{}) {
+					b.firstAccErr = time.Now()
+					logLevel = log.Debug
+				} else if time.Since(b.firstAccErr) < time.Minute {
+					logLevel = log.Debug
```

Comment:
The intention here was to only warn if it's been at least a minute since the accumulator errors started occurring. The code is a bit complex because we need to track when they started occurring, not when the last occurrence was. I'll have a look at refactoring it.